File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webapi/service/types/HistoryReadContext.java
package org.opcfoundation.webapi.service.types;

import org.eclipse.milo.opcua.stack.core.types.UaStructuredType;
import org.eclipse.milo.opcua.stack.core.types.enumerated.TimestampsToReturn;
import org.eclipse.milo.opcua.stack.core.types.structured.HistoryReadValueId;
import org.eclipse.milo.opcua.stack.core.types.structured.RequestHeader;
import org.springframework.lang.Nullable;
import org.springframework.web.context.request.NativeWebRequest;

import java.util.List;

public class HistoryReadContext extends ServiceContext {
    final private List<HistoryReadValueId> nodesToRead;
    final private UaStructuredType historyReadDetails;
    final private Boolean releaseContinuationPoints;
    final private TimestampsToReturn timestampsToReturn;

    public HistoryReadContext(
            List<HistoryReadValueId> nodesToRead,
            UaStructuredType historyReadDetails,
            Boolean releaseContinuationPoints,
            TimestampsToReturn timestampsToReturn,
            RequestHeader header,
            @Nullable String serverUri,
            NativeWebRequest webRequest)
    {
        super(header,serverUri,webRequest);
        this.nodesToRead = nodesToRead;
        this.historyReadDetails = historyReadDetails;
        this.releaseContinuationPoints = releaseContinuationPoints;
        this.timestampsToReturn = timestampsToReturn;
    }

    public List<HistoryReadValueId> getNodesToRead() {
        return nodesToRead;
    }

    public UaStructuredType getHistoryReadDetails() {
        return historyReadDetails;
    }

    public Boolean getReleaseContinuationPoints() {
        return releaseContinuationPoints;
    }

    public TimestampsToReturn getTimestampsToReturn() {
        return timestampsToReturn;
    }
}

File: opcua-webservice-node/src/server/types/contexts/HistoryReadContext.ts
import { ExtensionObject, HistoryReadRequest, HistoryReadValueId } from "opcua-webapi";
import { ServiceContext } from "./ServiceContext";

export class HistoryReadContext extends ServiceContext
{
    private _nodesToRead: Array<HistoryReadValueId>;
    private _historyReadDetails: ExtensionObject | null;
    private _timestampsToReturn: number;
    private _releaseContinuationPoints: boolean;   
    
    constructor(request: HistoryReadRequest, serverUri?: string)
    {
        super(serverUri, request.RequestHeader);
        this._nodesToRead = request.NodesToRead || [];
        this._historyReadDetails = (request.HistoryReadDetails) ? request.HistoryReadDetails : null;
        this._timestampsToReturn = request.TimestampsToReturn || 0;
        this._releaseContinuationPoints = request.ReleaseContinuationPoints || false;
    }

    get nodesToRead() : Array<HistoryReadValueId>
    {
        return this._nodesToRead;
    }

    get historyReadDetails() : ExtensionObject | null
    {
        return this._historyReadDetails;
    }

    get timestampsToReturn() : number
    {
        return this._timestampsToReturn;
    }

    get releaseContinuationPoints() : boolean
    {
        return this._releaseContinuationPoints;
    }
}